File: exercism/python/robot-simulator/robot_simulator.py
NORTH = 0
EAST = 1
SOUTH = 2
WEST = 3

DISPLACEMENTS = ((0, 1), (1, 0), (0, -1), (-1, 0))


class Robot(object):
    def __init__(self, bearing=NORTH, x=0, y=0):
        self.bearing = bearing
        self.coordinates = (x, y)

    def _turn(self, direction):
        self.bearing = (self.bearing + direction + 4) % 4

    def turn_left(self):
        self._turn(-1)

    def turn_right(self):
        self._turn(1)

    def advance(self):
        coord_disp = zip(self.coordinates, DISPLACEMENTS[self.bearing])
        self.coordinates = tuple(map(sum, coord_disp))

    def simulate(self, instructions):
        do = {'L': self.turn_left, 'R': self.turn_right, 'A': self.advance}
        [do[ch]() for ch in instructions]

File: exercism/javascript/robot-simulator/robot-simulator.js

const bearings = 'north east south west'.split(' ')

const step = {
  north: [0, 1],
  east: [1, 0],
  south: [0, -1],
  west: [-1, 0],
}

const commands = {
  L: 'turnLeft',
  R: 'turnRight',
  A: 'advance',
}

class Robot {
  orient (bearing) {
    if (!new Set(bearings).has(bearing)) throw 'Invalid Robot Bearing'
    this.bearing = bearing
  }
  at (x, y) {
    this.coordinates = [x, y]
  }
  place ({x, y, direction}) {
    this.orient(direction)
    this.at(x, y)
  }
  turnRight () { this._turn(1) }
  turnLeft () { this._turn(-1) }
  _turn (delta) {
    const bi = bearings.indexOf(this.bearing)
    this.bearing = bearings[(4 + bi + delta) % 4]
  }
  advance () {
    const [dx, dy] = step[this.bearing]
    const [x, y] = this.coordinates
    this.coordinates = [x + dx, y + dy]
  }
  instructions (str) {
    return str.split('').map(ch => commands[ch])
  }
  evaluate (str) {
    this.instructions(str).forEach(instruction => this[instruction]())
  }
}

module.exports = Robot
